pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,250 @@
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []kvserverpb.TracedEntry {
+	// NB: This check is an optimization to handle the common case where there
+	// are no registered traces. numRegistered is not a strict count of the
+	// number of traces, however it is incremented before we store a trace in
+	// the map and decremented before we remove it.
+	if r.numRegistered.Load() == 0 {
+		return nil
+	}
+
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// Close will unregister all the traces and log them. It should be called if the
+// replica is destroyed as it is possible some traces were never closed.
+func (r *RaftTracer) Close() {
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		r.numRegistered.Add(-1)
+		r.m.Delete(index)
+		t.logf(2, "cleanup log index (i%d) during close", index)
+		return true
+	})
+}
+
+// traceIfCovered will log the message if it touches any of the registered trace
+// points. Additionally it returns any saved contexts by index for sending to
+// remote nodes. This typically applies to messages that the leader sends to the
+// followers.
+func (r *RaftTracer) traceIfCovered(m raftpb.Message) []kvserverpb.TracedEntry {
+	var tracedEntries []kvserverpb.TracedEntry
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		if inRange(index, m.Entries) {
+			tracedEntries = append(tracedEntries, t.toTracedEntry(index))
+			// TODO(baptist): Not all the fields are relevant to log for all
+			// message types. Consider cleaning up what is logged.
+			t.logf(4,
+				"%s->%s %v Term:%d Log:%d/%d",
+				raft.DescribeTarget(m.From),
+				raft.DescribeTarget(m.To),
+				m.Type,
+				m.Term,
+				m.LogTerm,
+				m.Index,
+			)
+		}
+		return true
+	})
+	return tracedEntries
+}
+
+// inRange returns true if the index is within the range of the entries. It
+// assumes the entries are sorted by index. Specifically it returns if the index
+// is in the set [first_index,last_index].
+func inRange(index kvpb.RaftIndex, entries []raftpb.Entry) bool {
+	return len(entries) != 0 &&
+		index >= kvpb.RaftIndex(entries[0].Index) &&
+		index <= kvpb.RaftIndex(entries[len(entries)-1].Index)
+}
+
+// traceIfPast will log the message the message is past any registered tracing
+// points. It will additionally unregister traces that are no longer useful.
+// This typically applies to messages that followers send back to a leader.
+func (r *RaftTracer) traceIfPast(m raftpb.Message) {
+	if m.Reject {
+		return
+	}
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		shouldLog := false
+		shouldUnregister := false
+		switch m.Type {
+		case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp:
+			shouldLog = kvpb.RaftIndex(m.Index) >= index
+		case raftpb.MsgStorageApplyResp:
```

Comment:
That is a valid risk. There are a few options to deal with it, but none of them are minor. 

1) Reintroduce the term and logmark. That's not great as I'm still not sure we conclusively know when to clear this entry. 
2) Reintroduce the term and close out all traces if the term on the replica ever changes. I'm not sure if we will miss some important traces
3) Allow the tracing to stay active. While it continues to use the concurrent limit, it doesn't cause too much logging.

With the change to set the default to 0 and the protection on duplicate MsgAppResp and MsgStorageAppendResp I think this should be safe enough. The worse case is that we will lose logging ability rather than having too many traces or log messages active. I could also consider adding a "max trace time" in the future to expire traces that have not moved in too long, but I'm not sure how important that is to do.